security/pkg/k8s/controller/workloadsecret.go

```diff
@@ -488,37 +493,44 @@ func (sc *SecretController) scrtUpdated(oldObj, newObj interface{}) {
-		caSecret, scrtErr := sc.caSecretController.LoadCASecretWithRetry(CASecret,
-			sc.istioCaStorageNamespace, 100*time.Millisecond, 5*time.Second)
-		if scrtErr != nil {
-			k8sControllerLog.Errorf("Fail to load CA secret %s:%s (error: %s), skip updating secret %s",
-				sc.istioCaStorageNamespace, CASecret, scrtErr.Error(), name)
-			return
-		}
-		// The CA cert from istio-ca-secret is the source of truth. If CA cert
-		// in local keycertbundle does not match the CA cert in istio-ca-secret,
-		// reload root cert into keycertbundle.
-		if !bytes.Equal(caCert, caSecret.Data[caCertID]) {
-			k8sControllerLog.Warn("CA cert in KeyCertBundle does not match CA cert in " +
-				"istio-ca-secret. Start to reload root cert into KeyCertBundle")
-			var err error
-			rootCertificate, err = util.AppendRootCerts(caSecret.Data[caCertID], sc.rootCertFile)
-			if err != nil {
-				k8sControllerLog.Errorf("failed to append root certificates: %s", err.Error())
+	// Add a 30 seconds interval for key cert bundle sync to prevent I/O burst.
+	if sc.lastKCBSyncTime.IsZero() || time.Since(sc.lastKCBSyncTime) > 30*time.Second {
+		if !bytes.Equal(rootCertificate, scrt.Data[RootCertID]) {
+			caSecret, scrtErr := sc.caSecretController.LoadCASecretWithRetry(CASecret,
+				sc.istioCaStorageNamespace, 100*time.Millisecond, 5*time.Second)
+			if scrtErr != nil {
+				k8sControllerLog.Errorf("Fail to load CA secret %s:%s (error: %s), skip updating secret %s",
+					sc.istioCaStorageNamespace, CASecret, scrtErr.Error(), name)
 				return
 			}
-			if err := sc.ca.GetCAKeyCertBundle().VerifyAndSetAll(caSecret.Data[caCertID],
-				caSecret.Data[caPrivateKeyID], nil, rootCertificate); err != nil {
-				k8sControllerLog.Errorf("failed to reload root cert into KeyCertBundle (%v)", err)
-				return
+			// The CA cert from istio-ca-secret is the source of truth. If CA cert
+			// in local keycertbundle does not match the CA cert in istio-ca-secret,
+			// reload root cert into keycertbundle.
+			if !bytes.Equal(caCertInMem, caSecret.Data[caCertID]) {
+				k8sControllerLog.Warn("CA cert in KeyCertBundle does not match CA cert in " +
+					"istio-ca-secret. Start to reload root cert into KeyCertBundle")
+				var err error
+				rootCertificate, err = util.AppendRootCerts(caSecret.Data[caCertID], sc.rootCertFile)
+				if err != nil {
+					k8sControllerLog.Errorf("failed to append root certificates: %s", err.Error())
+					return
+				}
+				if err := sc.ca.GetCAKeyCertBundle().VerifyAndSetAll(caSecret.Data[caCertID],
+					caSecret.Data[caPrivateKeyID], nil, rootCertificate); err != nil {
+					k8sControllerLog.Errorf("failed to reload root cert into KeyCertBundle (%v)", err)
+					return
+				}
+				k8sControllerLog.Info("Successfully reloaded root cert into KeyCertBundle.")
+				sc.lastKCBSyncTime = time.Now()
+			} else {
+				log.Info("CA cert in KeyCertBundle matches CA cert in " +
+					"istio-ca-secret. Skip reloading root cert into KeyCertBundle")
 			}
```

Comment:
Please put this with the timestamp check into a separate function, like "tryToSyncKeyCertBundle".

Reply:
Good point. Done. Thanks!
The istio.io/istio/security/pkg/k8s/chiron takes about 74s to finish. We may want to reduce the time. 

?   	istio.io/istio/security/pkg/adapter/vault	[no test files]
ok  	istio.io/istio/security/pkg/caclient	1.768s
ok  	istio.io/istio/security/pkg/caclient/protocol	1.156s
?   	istio.io/istio/security/pkg/caclient/protocol/mock	[no test files]
ok  	istio.io/istio/security/pkg/cmd	0.077s
?   	istio.io/istio/security/pkg/credential	[no test files]
ok  	istio.io/istio/security/pkg/k8s/chiron	73.905s
ok  	istio.io/istio/security/pkg/k8s/configmap	0.142s
ok  	istio.io/istio/security/pkg/k8s/controller	3.921s
ok  	istio.io/istio/security/pkg/k8s/secret	0.135s
ok  	istio.io/istio/security/pkg/k8s/tokenreview	0.080s
ok  	istio.io/istio/security/pkg/listwatch	0.077s
ok  	istio.io/istio/security/pkg/nodeagent/cache	8.607s
?   	istio.io/istio/security/pkg/nodeagent/cache/mock	[no test files]
ok  	istio.io/istio/security/pkg/nodeagent/caclient	2.227s
?   	istio.io/istio/security/pkg/nodeagent/caclient/interface	[no test files]
ok  	istio.io/istio/security/pkg/nodeagent/caclient/providers/citadel	3.078s
ok  	istio.io/istio/security/pkg/nodeagent/caclient/providers/google	3.259s
ok  	istio.io/istio/security/pkg/nodeagent/caclient/providers/vault	0.191s
?   	istio.io/istio/security/pkg/nodeagent/model	[no test files]
?   	istio.io/istio/security/pkg/nodeagent/plugin	[no test files]
ok  	istio.io/istio/security/pkg/nodeagent/plugin/providers/google/stsclient	1.114s
ok  	istio.io/istio/security/pkg/nodeagent/sds	17.908s
ok  	istio.io/istio/security/pkg/nodeagent/secretfetcher	0.700s
?   	istio.io/istio/security/pkg/nodeagent/util	[no test files]
ok  	istio.io/istio/security/pkg/nodeagent/vm	0.368s
ok  	istio.io/istio/security/pkg/pki/ca	4.666s
?   	istio.io/istio/security/pkg/pki/ca/mock	[no test files]
ok  	istio.io/istio/security/pkg/pki/error	0.021s
ok  	istio.io/istio/security/pkg/pki/util	1.015s
?   	istio.io/istio/security/pkg/pki/util/mock	[no test files]
ok  	istio.io/istio/security/pkg/platform	0.435s
?   	istio.io/istio/security/pkg/platform/mock	[no test files]
ok  	istio.io/istio/security/pkg/probe	1.067s
ok  	istio.io/istio/security/pkg/registry	0.060s
ok  	istio.io/istio/security/pkg/registry/kube	0.139s
ok  	istio.io/istio/security/pkg/server/ca	6.977s
ok  	istio.io/istio/security/pkg/server/ca/authenticate	0.077s
?   	istio.io/istio/security/pkg/server/monitoring	[no test files]
?   	istio.io/istio/security/pkg/testing/sdsc	[no test files]
ok  	istio.io/istio/security/pkg/util	0.055s
?   	istio.io/istio/security/pkg/util/mock	[no test files]